### ISSUE TITLE
First and last name from OIDC is set to username

### ISSUE THREAD

**Issue by Natanande** [goharbor/harbor#13020]
During automatic onboarding using OIDC, the Realname is set to the username as seen [here](https://github.com/goharbor/harbor/blob/master/src/core/controllers/oidc.go#L198).

At my company, we set the username claim to "email", because Harbor requires the username to be unique. This results in the profile showing the email as firstname and last name, which doesn't make sense.

It would be nice to be able to set the Realname attribute to a specified claim (or a combination of multiple ones? given_name + family_name), the same way as we now can [with the username and group](https://github.com/goharbor/harbor/blob/master/src/common/utils/oidc/helper.go#L313).

This is mostly cosmetic, but it does look ridiculous when all three fields (username, email and realname) in the profile all are set to the user's email.

(Even better would be to dynamically update the realname at login/some interval, since that is something that can change)

**Comment by steven-zou** [comment]
cc @reasonerjt

**Comment by reasonerjt** [comment]
IMO to fix this we may allow user to set more attributes in the onboard dialog, which is a common practice.

In addition to that we may consider allow user to update his profile partially after he's onboarded.

**Comment by Natanande** [comment]
Agreed, but since we use the "Automatic onboarding" without any onboard dialog, we would prefer specifying a claim in the oidc JWT to automatically populate the profile. For us all user information is already stored centrally in Azure.

**Comment by kautkata** [comment]
I stumbled on this issue while trying to set up following authentication/authorization chain Harbor-2.1.1 -> Keycloak-10.0.1 -> MS AD.

From my perspective it would be enough to use value of "name" attribute returned by userinfo API.

**Comment by schuerg** [comment]
The OpenID Connect specification defines a set of standard claims for things like the first name, last name and so on.  
See: <https://openid.net/specs/openid-connect-basic-1_0.html#StandardClaims>

I think these claims (`preferred_username`, `given_name`, `family_name`) should be used if no other configuration is proivided to harbor.

**Comment by keliansb** [comment]
Issue still relevant, please do not close.

**Comment by lusu007** [comment]
Any news on this? In my opinion, it would be quite helpful if the `First and Lastname` field could be set automatically during onboarding

**Comment by Craftoncu** [comment]
Still a problem

**Comment by belfhi** [comment]
We also ran into this issue when configuring Keycloak as an OIDC Provider. Not a real problem but weird, especially since the Name is known and used as the username if `Username Claim` is not set.

**Comment by pniebylski-zilch** [comment]
Yeah, its still a problem

**Comment by NeckBeardPrince** [comment]
Incredible this hasn't been resolved yet

**Comment by shawnweeks** [comment]
Still not resolved in 2026

**Comment by NeckBeardPrince** [comment]
See you guys next year

**Comment by levi-pl** [comment]
I think it is time to look into the code :-)
